File: ViewControllerTesterTests/ViewControllerTestCase.m
//
//  ViewControllerTestCase.m
//  ViewControllerTester
//
//  Created by Randall Brown on 12/10/16.
//  Copyright © 2016 Randall Brown. All rights reserved.
//

#import "ViewControllerTestCase.h"
#import <ViewControllerTesterTests-Swift.h>


@implementation ViewControllerTestCase

-(void)spec {
    
}
    
+ (void)initialize {
    ViewControllerTestCase *spec = [self new];
    [VCTestHolder shared].currentClass = NSStringFromClass(self);
    [spec spec];
}
    
-(void)runTest:(VCTest*)test {
    [test run];
}

+ (SEL)addAppearanceInstanceMethodForTest:(VCTest *)test {
    IMP implementation = imp_implementationWithBlock(^(ViewControllerTestCase *testCase){
        NSLog(@"View Appears");
        [UIApplication sharedApplication].delegate.window.rootViewController = [VCTestHolder shared].currentViewController;
        
        [test run];
    });
    
    const char *types = [[NSString stringWithFormat:@"%s%s%s", @encode(id), @encode(id), @encode(SEL)] UTF8String];
    NSString *selectorName = [test.name stringByReplacingOccurrencesOfString:@" " withString:@"_"];
    SEL selector = NSSelectorFromString(selectorName);
    class_addMethod(self, selector, implementation, types);
    
    
    return selector;
}
    
+ (SEL)addDisappearanceInstanceMethodForTest:(VCTest *)test {
    IMP implementation = imp_implementationWithBlock(^(ViewControllerTestCase *testCase){
        [UIApplication sharedApplication].delegate.window.rootViewController = [[UIViewController alloc] init];
        
        [test run];
    });
    
    const char *types = [[NSString stringWithFormat:@"%s%s%s", @encode(id), @encode(id), @encode(SEL)] UTF8String];
    NSString *selectorName = [test.name stringByReplacingOccurrencesOfString:@" " withString:@"_"];
    SEL selector = NSSelectorFromString(selectorName);
    class_addMethod(self, selector, implementation, types);
    
    
    return selector;
}

+ (SEL)addInstanceMethodForTest:(VCTest *)test {
    IMP implementation = imp_implementationWithBlock(^(ViewControllerTestCase *testCase){
        [test run];
    });
    
    const char *types = [[NSString stringWithFormat:@"%s%s%s", @encode(id), @encode(id), @encode(SEL)] UTF8String];
    NSString *selectorName = [test.name stringByReplacingOccurrencesOfString:@" " withString:@"_"];
    SEL selector = NSSelectorFromString(selectorName);
    class_addMethod(self, selector, implementation, types);

    
    return selector;
}
    
+ (NSArray*)testInvocations {
    [VCTestHolder shared].currentClass = NSStringFromClass(self);
    if([VCTestHolder shared].appearanceTests.count + [VCTestHolder shared].disappearanceTests.count <= 0) {
        return nil;
    }

    NSMutableArray *invocations = [NSMutableArray array];
    
    for(VCTest *test in [VCTestHolder shared].appearanceTests) {
        
        SEL selector;
        if(test == [VCTestHolder shared].appearanceTests.firstObject){
            selector = [self addAppearanceInstanceMethodForTest:test];
        }
        else {
            selector = [self addInstanceMethodForTest:test];
        }
        
        NSMethodSignature *methodSignature = [[self class] instanceMethodSignatureForSelector:selector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:methodSignature];
        invocation.selector = selector;
        
        [invocations addObject:invocation];
    }
    
    for(VCTest *test in [VCTestHolder shared].disappearanceTests) {
        SEL selector;
        if(test == [VCTestHolder shared].disappearanceTests.firstObject){
            selector = [self addDisappearanceInstanceMethodForTest:test];
        }
        else {
            selector = [self addInstanceMethodForTest:test];
        }
        
        NSMethodSignature *methodSignature = [[self class] instanceMethodSignatureForSelector:selector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:methodSignature];
        invocation.selector = selector;
        
        [invocations addObject:invocation];
    }
    
    return invocations;
}
    
    - (void)recordFailureWithDescription:(NSString *)description
                                  inFile:(NSString *)filePath
                                  atLine:(NSUInteger)lineNumber
                                expected:(BOOL)expected {
        
        [self.testRun recordFailureWithDescription:description
                                                   inFile:filePath
                                                   atLine:lineNumber
                                                 expected:expected];
    }
    
    
@end

File: ViewControllerTesterTests/ViewControllerTestCase.swift
import XCTest
import Nimble
import Quick

public class VCTest: NSObject {
    init(name: String, _ test: @escaping ()->()) {
        self.test = test
        self.name = name
    }
    
    func run() {
        test()
    }
    
    private var test: ()->()
    var name: String
}

func vc(_ description: String, _ test:@escaping ()->()) {
    VCTestHolder.shared.add(test: VCTest(name:description, test))
}

func viewAppears(_ test:@escaping ()->()) {
    VCTestHolder.shared.addAppear()
    test()
}

func viewDisappears(_ test:@escaping ()->()) {
    VCTestHolder.shared.addDisappear()
    test()
}

func testViewController(_ viewController: UIViewController, _ test: ()->()) {
    VCTestHolder.shared.currentViewController = viewController
    test()
}


//class ViewControllerTestCase2: QuickSpec {
//    private var appears: (() -> ())?
//    private var disappears: (() -> ())?
//    
//    func testViewController(_ viewController: UIViewController, test: () -> ()) {
//        //let className = String(describing: Mirror(reflecting: viewController).subjectType)
//        
//        test()
//        
//        UIApplication.shared.delegate?.window??.rootViewController = viewController
//        expect(true).toEventually(beTrue())
//        
//        appears?()
//        
//        
//        UIApplication.shared.delegate?.window??.rootViewController = UIViewController()
//        expect(true).toEventually(beTrue())
//        
//        disappears?()
//    }
//    
//    func viewAppears(_ appear: @escaping () -> ()) {
//        appears = appear
//        
//    }
//    
//    func test(_ description: String, _ test: () throws -> ()) {
//        do {
//            try test()
//        } catch {
//            print("Failed")
//        }
//        
//    }
//    
//    func viewDisappears( _ disappear: @escaping () -> ()) {
//        disappears = disappear
//    }
//}
